### ISSUE TITLE
Wrong result when simd-ing over BitArrays

### ISSUE THREAD

**Issue by mbauman** [JuliaLang/julia#27482]
```julia
julia> any!(falses(10), trues(10, 10))
10-element BitArray{1}:
 false
 false
 false
 false
 false
 false
 false
  true
  true
  true

julia> any!(Vector(falses(10)), trues(10, 10))
10-element Array{Bool,1}:
 true
 true
 true
 true
 true
 true
 true
 true
 true
 true
```

I'm guessing this is an AVX512 codegen issue with the new compiler, but #27182 does not fix it.  Doesn't occur with a 44-day old build (9f5351c36a) which is before the new compiler was enabled.

**Comment by mbauman** [comment]
Minimal example:

```julia
function g(f, op, R::AbstractArray, A::AbstractArray)
    @inbounds for IA in axes(A, 2)
        @simd for i in axes(A, 1)
            R[i] = op(R[i], f(A[i,IA]))
        end
    end
    return R
end

g(identity, |, falses(10), trues(10, 10))
```

**Comment by Keno** [comment]
Yeah, I just got to the same reduction. That `@simd` is illegal, because the memory accesses are not loop-parallel.

**Comment by mbauman** [comment]
Oh, interesting.  This restriction makes sense, but it's news to me.  So we cannot `@simd` over assignments to BitArrays because they affect the same `UInt64` value across multiple iterations?

**Comment by Keno** [comment]
> So we cannot @simd over assignments to BitArrays because they affect the same UInt64 value across multiple iterations?

Correct. Technically not over reads from BitArrays either.

**Comment by mbauman** [comment]
Dang.  That really hamstrings our ability to use `@simd` in generic array code.  Looks like we're doing it wrong in reductions, `extrema!`, and `broadcast`.

I'm guessing lots of packages will hit this, too.  I'd really love it if we could just fail to vectorize in cases like this instead of returning wrong results.

**Comment by Keno** [comment]
> I'm guessing lots of packages will hit this, too. I'd really love it if we could just fail to vectorize in cases like this instead of returning wrong results.

Remove `@simd` ;)

**Comment by mbauman** [comment]
Just for what it's worth — yesterday before we had totally figured this out I had kicked off a bisect.  You probably already know this, but it was 56d7ebeed2ec7edbd3e1b8c5315561b650a65d67 that started causing the incorrect results.

**Comment by StefanKarpinski** [comment]
For what it's worth, I don't think our compiler "hints" should ever have license to produce incorrect results like this. I understand that this is a bug; documenting the bug is fine for now, but at some point this does need fixing.

**Comment by Keno** [comment]
That's literally the only thing that there annotations are for. Giving the compiler license to do transformations that it otherwise thinks are illegal. If the compiler though it was allowed to do this, it'd have done it iself.

**Comment by StefanKarpinski** [comment]
Typical uses of `@simd` give the compiler license to re-associate floating-point operations even though that technically changes the result, although typically not by much. This does not seem like the same class of license.

**Comment by vtjnash** [comment]
Keno's definitely correct. Could we do something like having an `@simd_unsafe` primitive though that we use to mark the BitArray getters and setters? I suspect that bitarray is fairly unique in its violation of  memory reference expectations.

**Comment by StefanKarpinski** [comment]
That's a fair point that bit arrays are a bit unusual in this regard.

**Comment by mbauman** [comment]
Sparse arrays would also incur trouble here.  And even structured arrays won't have a consistent read memory access pattern — does that mean you have liberty to give me garbage?  Or does that just mean it'll fail to vectorize?

**Comment by vtjnash** [comment]
The guarantee we appear to be making is that each iteration of the loop will not observe the memory write from any other loop iteration. Random access pattern shouldn't be a problem (it'll likely just not be possible to vectorize). The BitArray issue is that a write actually consists of a read-modify-write (RMW) that makes it appear race-y here.

**Comment by KristofferC** [comment]
So we have to remove simd annotation on all generic code?

**Comment by mbauman** [comment]
Definitely generic `setindex!` is not safe.

**Comment by oscardssmith** [comment]
Instead could we make `@simd` do nothing for `bitarrays` and have a special function when you want `simd` for them? That seems better than breaking the ability to write fast generic code.

**Comment by saolof** [comment]
This does remind me of some errors related to the vector<bool> specialization in C++. If you want to expose a safe generic interface, breaking assumptions related to memory packing like bitarray does definitely breaks that.

Maybe add a macro to mark a method like setindex! for BitArray as thread-unsafe to disable SIMD-optimizations on code that calls it?

**Comment by Keno** [comment]
This bug makes Pkg3 completely useless on Skylake machines. We know from bugreports that we have a significant numbers of users on such machines. Bumping this for the next 0.7 pre-release.

**Comment by StefanKarpinski** [comment]
Yeah, this is definitely a priority to get a fix in for.

**Comment by chethega** [comment]
In some sense, the problem is not that BitArray get/set are bad for SIMD, it is that they are too good for SIMD. That is, BitArrays getindex/setindex should talk to the simd machinery in order to try to get data processed in 64 bit chunks (or even better 128/256/512) instead of dropping down to 1-bit sequential processing.

**Comment by mbauman** [comment]
It's not just `BitArray`, though.  We were allowing arbitrary functions inside `@simd` loops — functions which could have similar sorts of side-effects that have a catastrophic interaction with SIMD.

**Comment by Keno** [comment]
Fixed by #27670
